File: sam3/ios/Sources/Sam3Detector.swift
import CoreGraphics
import Foundation
import UIKit

/// SAM 3.1 text-prompted detection + instance segmentation (image side of the
/// Object-Multiplex checkpoint) on LiteRT CompiledModel, three graphs:
///   vision : image [1,3,1008,1008] -> [fpn288 | fpn144 | fpn72]         (Metal fp16)
///   text   : token embeddings [1,32,1024] -> text memory [32*256]       (CPU — exact;
///            the CLIP-L residual stream reaches |x|~1.2e3 and fp16 GPU execution
///            corrupts the prompt embedding for some prompts)
///   head   : [fpn x3 | text_mem | pad(32)] -> [logits(200) | boxes(200x4 cxcywh,
///            normalized) | presence(1) | mask logits (200x288x288)]     (Metal f32)
/// Host: BPE tokenize, fp16 token-embedding lookup, score threshold, mask sigmoid.
/// score = sigmoid(logit) * sigmoid(presence); keep > threshold.
final class Sam3Detector {
    static let size = 1008
    static let tok = 32
    static let tdim = 1024
    static let queries = 200
    static let maskSize = 288
    static let nVis = 256 * (288 * 288 + 144 * 144 + 72 * 72)

    struct Detection: Identifiable {
        let id = UUID()
        let score: Float
        /// cx, cy, w, h normalized to [0,1]
        let box: [Float]
        /// sigmoid mask probabilities, maskSize x maskSize, row-major
        let mask: [Float]
    }

    struct StageTiming {
        var visionMs = 0.0
        var textMs = 0.0
        var headMs = 0.0
        var visionCached = false
    }

    /// Model files live in Documents (copied over from Finder / the Files app);
    /// bundle resources are the fallback so a dev build can also embed them.
    static func locate(_ name: String) -> URL? {
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let inDocs = docs.appendingPathComponent(name)
        if FileManager.default.fileExists(atPath: inDocs.path) { return inDocs }
        let stem = (name as NSString).deletingPathExtension
        let ext = (name as NSString).pathExtension
        return Bundle.main.url(forResource: stem, withExtension: ext)
    }

    static let requiredFiles = [
        "sam3_vision.tflite", "sam3_text.tflite", "sam3_head.tflite",
        "sam3_token_embed.bin", "vocab.json", "merges.txt",
    ]

    static func missingFiles() -> [String] {
        requiredFiles.filter { locate($0) == nil }
    }

    private let tokenizer: BpeTokenizer
    private let vision: Sam3Graph
    private let text: Sam3Graph
    private let head: Sam3Graph
    /// fp16 [49408 x 1024] row-major token-embedding table, memory-mapped (101 MB).
    private let tokenTable: Data

    private(set) var timing = StageTiming()
    var compileReport: String {
        [vision, text, head]
            .map { "\($0.name): \($0.accel.label), compile \(String(format: "%.1f", $0.compileSeconds))s, fully_accelerated=\($0.fullyAccelerated)" }
            .joined(separator: "\n")
    }

    /// Compiles all three graphs; slow on first launch (Metal shader compile).
    /// `progress` is called with a stage label before each compile.
    init(progress: @escaping (String) -> Void) throws {
        func url(_ name: String) throws -> URL {
            guard let u = Sam3Detector.locate(name) else {
                throw LiteRTError.interface("missing \(name)")
            }
            return u
        }
        // Compile with the preferred accelerator, falling back down the list so a
        // delegate failure on some device/OS degrades instead of blocking the app.
        func compile(_ name: String, prefer: [Sam3Accel]) throws -> Sam3Graph {
            let path = try url(name).path
            var lastError: Error?
            for accel in prefer {
                do {
                    return try Sam3Graph(path: path, accel: accel)
                } catch {
                    lastError = error
                    print("SAM3: \(name) on \(accel.label) failed: \(error)")
                }
            }
            throw lastError ?? LiteRTError.interface("no accelerator for \(name)")
        }
        progress("Loading tokenizer…")
        tokenizer = try BpeTokenizer(vocabURL: try url("vocab.json"), mergesURL: try url("merges.txt"))
        tokenTable = try Data(contentsOf: try url("sam3_token_embed.bin"), options: .alwaysMapped)
        progress("Compiling vision graph on the GPU…")
        vision = try compile("sam3_vision.tflite", prefer: [.gpu, .cpu])
        progress("Compiling text graph…")
        text = try compile("sam3_text.tflite", prefer: [.cpu])
        progress("Compiling detection head on the GPU…")
        head = try compile("sam3_head.tflite", prefer: [.gpuF32, .gpu, .cpu])
    }

    var visionAccel: String { vision.accel.label }
    var textAccel: String { text.accel.label }
    var headAccel: String { head.accel.label }

    private func preprocess(_ image: UIImage) -> [Float] {
        let side = Self.size
        let space = CGColorSpaceCreateDeviceRGB()
        var pixels = [UInt8](repeating: 0, count: side * side * 4)
        pixels.withUnsafeMutableBytes { raw in
            let ctx = CGContext(
                data: raw.baseAddress, width: side, height: side, bitsPerComponent: 8,
                bytesPerRow: side * 4, space: space,
                bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue)!
            ctx.interpolationQuality = .medium
            if let cg = image.cgImage {
                ctx.draw(cg, in: CGRect(x: 0, y: 0, width: side, height: side))
            }
        }
        let plane = side * side
        var out = [Float](repeating: 0, count: 3 * plane)
        out.withUnsafeMutableBufferPointer { o in
            pixels.withUnsafeBufferPointer { p in
                for i in 0..<plane {
                    o[i] = (Float(p[i * 4]) / 255 - 0.5) / 0.5
                    o[plane + i] = (Float(p[i * 4 + 1]) / 255 - 0.5) / 0.5
                    o[2 * plane + i] = (Float(p[i * 4 + 2]) / 255 - 0.5) / 0.5
                }
            }
        }
        return out
    }

    // Vision features are per-image; cache them so re-prompting skips the big graph.
    private var cachedKey: ObjectIdentifier?
    private var visFeat: [Float] = []

    private func runVision(_ image: UIImage) throws {
        if cachedKey == ObjectIdentifier(image) {
            timing.visionMs = 0
            timing.visionCached = true
            return
        }
        let t0 = Date()
        let input = preprocess(image)
        visFeat = try vision.run(input)
        timing.visionMs = Date().timeIntervalSince(t0) * 1000
        timing.visionCached = false
        cachedKey = ObjectIdentifier(image)
    }

    private func runText(_ prompt: String) throws -> (mem: [Float], pad: [Float]) {
        let t0 = Date()
        let ids = tokenizer.encode(prompt)
        var emb = [Float](repeating: 0, count: Self.tok * Self.tdim)
        tokenTable.withUnsafeBytes { (raw: UnsafeRawBufferPointer) in
            let table = raw.bindMemory(to: Float16.self)
            for t in 0..<Self.tok {
                let base = ids[t] * Self.tdim
                for d in 0..<Self.tdim {
                    emb[t * Self.tdim + d] = Float(table[base + d])
                }
            }
        }
        let mem = try text.run(emb)  // [32*256]
        let pad = (0..<Self.tok).map { ids[$0] == 0 ? Float(1) : Float(0) }
        timing.textMs = Date().timeIntervalSince(t0) * 1000
        return (mem, pad)
    }

    /// Returns detections above `threshold`, unsorted (query order).
    func detect(image: UIImage, prompt: String, threshold: Float = 0.5) throws -> [Detection] {
        try runVision(image)
        let (mem, pad) = try runText(prompt)
        let t0 = Date()
        var headInput = [Float](repeating: 0, count: Self.nVis + Self.tok * 256 + Self.tok)
        headInput.replaceSubrange(0..<Self.nVis, with: visFeat)
        headInput.replaceSubrange(Self.nVis..<(Self.nVis + Self.tok * 256), with: mem)
        headInput.replaceSubrange((Self.nVis + Self.tok * 256)..., with: pad)
        let y = try head.run(headInput)
        timing.headMs = Date().timeIntervalSince(t0) * 1000

        let presence = 1 / (1 + exp(-y[1000]))
        var out: [Detection] = []
        let maskArea = Self.maskSize * Self.maskSize
        for q in 0..<Self.queries {
            let score = 1 / (1 + exp(-y[q])) * presence
            if score <= threshold { continue }
            let box = Array(y[(200 + q * 4)..<(200 + q * 4 + 4)])
            let base = 1001 + q * maskArea
            var mask = [Float](repeating: 0, count: maskArea)
            for i in 0..<maskArea { mask[i] = 1 / (1 + exp(-y[base + i])) }
            out.append(Detection(score: score, box: box, mask: mask))
        }
        return out
    }
}

File: sam3/app/src/main/java/com/sam3/Sam3Detector.kt
package com.sam3

import android.content.Context
import android.graphics.Bitmap
import com.google.ai.edge.litert.Accelerator
import com.google.ai.edge.litert.CompiledModel
import java.io.Closeable
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * SAM 3.1 text-prompted detection + instance segmentation (image side of the
 * Object-Multiplex checkpoint) on the LiteRT CompiledModel API, three graphs:
 *   vision : image [1,3,1008,1008] -> [fpn288 | fpn144 | fpn72]           (GPU)
 *   text   : token embeddings [1,32,1024] -> text memory [1, 32*256]      (CPU — the
 *            CLIP-L residual stream reaches |x|~1.2e3; fp16 GPU execution corrupts the
 *            prompt embedding for some prompts, and this graph is fast anyway)
 *   head   : [fpn x3 | text_mem | pad(32)] -> [logits(200) | boxes(200x4 cxcywh,
 *            normalized) | presence(1) | mask logits (200x288x288)]       (GPU)
 * Host: BPE tokenize, fp16 token-embedding lookup, score threshold, mask sigmoid.
 * score = sigmoid(logit) * sigmoid(presence); keep > threshold.
 */
class Sam3Detector(private val ctx: Context) : Closeable {

    companion object {
        const val SIZE = 1008
        const val TOK = 32
        const val TDIM = 1024
        const val QUERIES = 200
        const val MASK = 288
        const val N_VIS = 256 * (288 * 288 + 144 * 144 + 72 * 72)
    }

    data class Detection(
        val score: Float,
        /** cx, cy, w, h normalized to [0,1] */
        val box: FloatArray,
        /** sigmoid mask probabilities, MASK x MASK, row-major */
        val mask: FloatArray,
    )

    private fun f(name: String) = File(ctx.filesDir, name).also {
        check(it.exists()) { "Missing ${it.name}. Run scripts/install_to_device.sh first." }
    }

    private val tokenizer = BpeTokenizer(ctx)
    private val visModel = CompiledModel.create(f("sam3_vision.tflite").absolutePath,
        CompiledModel.Options(Accelerator.GPU), null)
    private val visIn = visModel.createInputBuffers()
    private val visOut = visModel.createOutputBuffers()
    private val textModel = CompiledModel.create(f("sam3_text.tflite").absolutePath,
        CompiledModel.Options(Accelerator.CPU), null)
    private val textIn = textModel.createInputBuffers()
    private val textOut = textModel.createOutputBuffers()
    private val headModel = CompiledModel.create(f("sam3_head.tflite").absolutePath,
        CompiledModel.Options(Accelerator.GPU), null)
    private val headIn = headModel.createInputBuffers()
    private val headOut = headModel.createOutputBuffers()

    /** fp16 [49408 x 1024] row-major token-embedding table, kept as raw bytes (101 MB). */
    private val tokEmb: ByteBuffer =
        ByteBuffer.wrap(f("sam3_token_embed.bin").readBytes()).order(ByteOrder.LITTLE_ENDIAN)

    var lastVisionMs = 0L; var lastTextMs = 0L; var lastHeadMs = 0L

    private fun halfToFloat(h: Int): Float {
        val s = (h ushr 15) and 0x1; val e = (h ushr 10) and 0x1F; val m = h and 0x3FF
        val bits = when {
            e == 0 -> if (m == 0) s shl 31 else {
                var mant = m; var exp = -1
                do { mant = mant shl 1; exp++ } while (mant and 0x400 == 0)
                (s shl 31) or ((127 - 15 - exp) shl 23) or ((mant and 0x3FF) shl 13)
            }
            e == 0x1F -> (s shl 31) or (0xFF shl 23) or (m shl 13)
            else -> (s shl 31) or ((e - 15 + 127) shl 23) or (m shl 13)
        }
        return Float.fromBits(bits)
    }

    private fun preprocess(bm: Bitmap): FloatArray {
        val s = Bitmap.createScaledBitmap(bm, SIZE, SIZE, true)
        val px = IntArray(SIZE * SIZE); s.getPixels(px, 0, SIZE, 0, 0, SIZE, SIZE)
        val out = FloatArray(3 * SIZE * SIZE)
        val plane = SIZE * SIZE
        for (i in px.indices) {
            val p = px[i]
            out[i] = (((p shr 16) and 0xFF) / 255f - 0.5f) / 0.5f
            out[plane + i] = (((p shr 8) and 0xFF) / 255f - 0.5f) / 0.5f
            out[2 * plane + i] = ((p and 0xFF) / 255f - 0.5f) / 0.5f
        }
        return out
    }

    // vision features are per-image; cache them so re-prompting skips the big graph
    private var cachedKey: Bitmap? = null
    lateinit var visFeat: FloatArray
    /** debug taps for the on-device parity check (autotest) */
    var lastTextMem: FloatArray? = null
    var lastPad: FloatArray? = null
    var lastHeadOut: FloatArray? = null
    var lastInput: FloatArray? = null

    private fun runVision(bm: Bitmap) {
        if (cachedKey === bm) { lastVisionMs = 0; return }
        val t0 = System.nanoTime()
        val inp = preprocess(bm)
        lastInput = inp
        visIn[0].writeFloat(inp)
        visModel.run(visIn, visOut)
        visFeat = visOut[0].readFloat()
        lastVisionMs = (System.nanoTime() - t0) / 1_000_000
        cachedKey = bm
    }

    private fun runText(prompt: String): Pair<FloatArray, FloatArray> {
        val t0 = System.nanoTime()
        val ids = tokenizer.encode(prompt)
        val emb = FloatArray(TOK * TDIM)
        for (t in 0 until TOK) {
            val base = ids[t] * TDIM * 2
            for (d in 0 until TDIM) {
                emb[t * TDIM + d] = halfToFloat(tokEmb.getShort(base + d * 2).toInt() and 0xFFFF)
            }
        }
        textIn[0].writeFloat(emb)
        textModel.run(textIn, textOut)
        val mem = textOut[0].readFloat()                        // [32*256]
        val pad = FloatArray(TOK) { if (ids[it] == 0) 1f else 0f }
        lastTextMem = mem; lastPad = pad
        lastTextMs = (System.nanoTime() - t0) / 1_000_000
        return Pair(mem, pad)
    }

    /** @return detections above [threshold], unsorted (query order). */
    fun detect(bm: Bitmap, prompt: String, threshold: Float = 0.5f): List<Detection> {
        runVision(bm)
        val (mem, pad) = runText(prompt)
        val t0 = System.nanoTime()
        val headInput = FloatArray(N_VIS + TOK * 256 + TOK)
        System.arraycopy(visFeat, 0, headInput, 0, N_VIS)
        System.arraycopy(mem, 0, headInput, N_VIS, TOK * 256)
        System.arraycopy(pad, 0, headInput, N_VIS + TOK * 256, TOK)
        headIn[0].writeFloat(headInput)
        headModel.run(headIn, headOut)
        val y = headOut[0].readFloat()
        lastHeadOut = y
        lastHeadMs = (System.nanoTime() - t0) / 1_000_000
        val presence = 1f / (1f + kotlin.math.exp(-y[1000]))
        val out = ArrayList<Detection>()
        for (q in 0 until QUERIES) {
            val score = 1f / (1f + kotlin.math.exp(-y[q])) * presence
            if (score <= threshold) continue
            val box = FloatArray(4) { y[200 + q * 4 + it] }
            val mask = FloatArray(MASK * MASK)
            val base = 1001 + q * MASK * MASK
            for (i in mask.indices) mask[i] = 1f / (1f + kotlin.math.exp(-y[base + i]))
            out.add(Detection(score, box, mask))
        }
        return out
    }

    override fun close() {
        visIn.forEach { it.close() }; visOut.forEach { it.close() }; visModel.close()
        textIn.forEach { it.close() }; textOut.forEach { it.close() }; textModel.close()
        headIn.forEach { it.close() }; headOut.forEach { it.close() }; headModel.close()
    }
}
